### ISSUE TITLE
Using from a single binary

### ISSUE THREAD

**Issue by nyurik** [asg017/sqlite-loadable-rs#12]
From a deployment/usage perspective, it would be ideal if the resulting executable could contain both the statically-linked sqlite, as well as any extensions. Do you see any way to achieve this?

**Comment by asg017** [comment]
What use case are you thinking of? 

`sqlite-loadable-rs` could be compiled with `crate-type = ["staticlib"]` (in addition to the typical `"dylib" crate type), which can then be used to statically compile a SQLite extension into an application. Typically the application (your CLI, webserver, etc.) with handle both statically linking whatever sqlite version they want, in addition to statically linking any sqlite extensions they like. 


I haven't blogged about it much, but it's possible to statically link a sqlite extension written with `sqlite-loadable-rs` into an application. For example the [`sqlite-regex`](https://crates.io/crates/sqlite-regex) crate can be added to your Rust projects, as [described here](https://github.com/asg017/sqlite-regex/issues/10#issuecomment-1585399669). You add rusqlite as a separate dependency, which can statically link sqlite. 

You can also use it to compile `.a` files, generate a `.h` file and use those to statically link in C/Go. I've done it before but haven't written much about it. 

Is that the kind of thing you're going for?

**Comment by nyurik** [comment]
@asg017 thank you so much for such a detailed answer!

I was thinking of doing this:
* develop a new crate that implements `md5` hashing function for sqlite
* in my app that uses sqlx, add the md5 extension crate and statically link it (together with the sqlite)
* my app will use `query("SELECT md5(foo) ...")` types of queries by loading needed extension at runtime. The sqlite DB files should not be modified just because I'm using an extra function - it should simply be an internal functionality of my app.
* ideally, i would love to use sqlx's compile-time statement validation via a macro:  `query!("SELECT md5(foo) ...")`
* most importantly, the app should remain just a single binary file that contains sqlite and md5 extension

**Comment by asg017** [comment]
Yeah, I think statically compiling an extension written with `sqlite-loadable-rs` will do what you'd like! The [`sqlite3_auto_extension `](https://www.sqlite.org/c3ref/auto_extension.html) function will probably do what you'd like - you'd call that with the entrypoint of your extension that defines `md5()`, and any time after that, any new SQLite connection will have the `md5()` function available. Since everything is statically linked, it'll be a single binary, no extra loadable files needed.  That `sqlite-regex` issue link above will be a good example to work off of. 

I'm not too sure about the `query!()` macro however - does that mean your SQL gets compiled + executed at build time? Not familiar with sqlx.

I'll probably blog about statically linking SQLite extension in the next few weeks, and will also include an end-to-end example of statically linking in this repo soon as well.

**Comment by asg017** [comment]
That being said `sqlite-loadable-rs` might be overkill if your custom SQL functions will _only_ be used in Rust. You're probably gonna use rusqlite anyway, and they have a [`create_scalar_function` function](https://docs.rs/rusqlite/latest/rusqlite/struct.Connection.html#method.create_scalar_function) that can implement a `md5()` function just fine. Unless you plan to also re-use your custom SQL functions in other languages (Python, Node.js, etc.), then using rusqlite/`create_scalar_function` will likely be simpler.

**Comment by nyurik** [comment]
@asg017 thank you for all the info!  I wrote [sqlite-hashes ](https://crates.io/crates/sqlite-hashes) crate for the static usacase using rusqlite, but I do hope that I can somehow support dynamic lib generation as well, e.g. with a feature flag of sorts.

**Comment by asg017** [comment]
Can you share some error messages that you see? Along with the platform you are compiling on (Macos/linux/arm etc.), the sqlite-loadable version you're on, and the compiler flags you've tried?

Also, does your Rust project have any additional dependencies that reply on other C libraries?

**Comment by anacrolix** [comment]
I think like @nyurik I haven't gone down the path of building on this crate. I have built on rusqlite. I believe the linker issues I'm having are not related to rusqlite. I'm currently making use of what is described in https://ricardoanderegg.com/posts/extending-sqlite-with-rust/ to create a dynamically loadable extension with rusqlite. I will be very interested to read your blog post if it talks about statically linking such an extension in to an executable. The branch of rusqlite referred to in the link has an example of statically linking an extension into a shared object that is loaded dynamically, it's not quite the same.

**Comment by anacrolix** [comment]
Here is the error if it means anything to you:
```
/Users/anacrolix/src/go1.21/pkg/tool/darwin_arm64/link: running clang++ failed: exit status 1
Undefined symbols for architecture arm64:
  "_sqlite3_api", referenced from:
      std::sync::once::Once::call_once::_$u7b$$u7b$closure$u7d$$u7d$::h1b8a0b94711ed9e1 in libdht_indexer.a(libsqlite3_sys-c5df21685f701ba8.libsqlite3_sys.61575ea2ee6fb45b-cgu.4.rcgu.o)
      core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::hea4cd6175a270b2a in libdht_indexer.a(libsqlite3_sys-c5df21685f701ba8.libsqlite3_sys.61575ea2ee6fb45b-cgu.4.rcgu.o)
      libsqlite3_sys::loadable_extension::loadable_extension_embedded_init::h57de3cfefb39c493 in libdht_indexer.a(libsqlite3_sys-c5df21685f701ba8.libsqlite3_sys.61575ea2ee6fb45b-cgu.4.rcgu.o)
     (maybe you meant: libsqlite3_sys::loadable_extension::loadable_extension_sqlite3_api::h128c8b56cbeba93d)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
